File: mobile/modules/keating-needle/android/src/main/java/com/diogenesoftoronto/keating/needle/KeatingNeedleModule.kt
package com.diogenesoftoronto.keating.needle

import android.net.Uri
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.nio.CharBuffer
import java.nio.charset.CodingErrorAction
import java.security.MessageDigest
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

internal object NeedleNative {
  val linked: Boolean = try { System.loadLibrary("keating_needle"); true } catch (_: UnsatisfiedLinkError) { false }
  @JvmStatic external fun supported(): Boolean
  @JvmStatic external fun revision(): String
  @JvmStatic external fun identity(): String?
  @JvmStatic external fun filename(): String
  @JvmStatic external fun modelSha(): String
  @JvmStatic external fun modelSize(): Int
  @JvmStatic external fun embed(model: ByteArray, texts: Array<ByteArray>): Array<FloatArray>
}

class KeatingNeedleModule : Module() {
  companion object {
    // Engine ownership is process-global, including multiple Expo module instances.
    private val worker = Executors.newSingleThreadExecutor()
    private val busy = AtomicBoolean(false)
  }
  @Volatile private var closed = false
  private fun supported() = NeedleNative.linked && NeedleNative.supported()
  private fun requireSupported() { check(supported()) { "Needle needs an ARM64 native Keating build." } }
  private fun directory(): File {
    requireSupported()
    val context = appContext.reactContext ?: error("Keating is not ready.")
    return File(context.noBackupFilesDir, "needle-retrieval").also { check(it.isDirectory || it.mkdirs()) }.canonicalFile
  }
  private fun modelFile(uri: String, partialAllowed: Boolean): File {
    val parsed = Uri.parse(uri)
    require(parsed.scheme == "file" && parsed.authority.isNullOrEmpty() && parsed.query == null && parsed.fragment == null) { "Needle needs its private local model file." }
    val requested = File(parsed.path ?: error("Invalid model URI"))
    val file = requested.canonicalFile
    require(file == requested.absoluteFile && file.parentFile == directory().canonicalFile && (file.name == NeedleNative.filename() || (partialAllowed && file.name == NeedleNative.filename() + ".part"))) { "Needle model path is outside its private storage." }
    return file
  }
  private fun verifiedBytes(file: File): ByteArray? {
    if (!file.isFile || file.length() != NeedleNative.modelSize().toLong()) return null
    // Hash the exact bytes passed to native; no hash-then-reopen race.
    val bytes = file.inputStream().use { input ->
      val result = ByteArray(NeedleNative.modelSize())
      var offset = 0
      while (offset < result.size) { val count = input.read(result, offset, result.size - offset); if (count <= 0) return null; offset += count }
      if (input.read() != -1) return null
      result
    }
    val digest = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }
    return if (digest == NeedleNative.modelSha()) bytes else null
  }
  override fun definition() = ModuleDefinition {
    Name("KeatingNeedle")
    Constants("supported" to supported(), "runtimeRevision" to if (NeedleNative.linked) NeedleNative.revision() else "",
      "modelIdentity" to if (supported()) NeedleNative.identity() else null)
    AsyncFunction("getModelDirectoryAsync") { Uri.fromFile(directory()).toString() }
    AsyncFunction("createModelFileAsync") {
      val file = File(directory(), NeedleNative.filename() + ".part")
      require(file.canonicalFile == file.absoluteFile && file.canonicalFile.parentFile == directory().canonicalFile) { "Invalid Needle download path." }
      file.createNewFile()
      Uri.fromFile(file).toString()
    }
    AsyncFunction("verifyModelFileAsync") { uri: String, promise: Promise ->
      if (closed || !busy.compareAndSet(false, true)) promise.reject("E_NEEDLE_BUSY", "Needle is busy or unavailable.", null)
      else worker.execute {
        try {
          val verified = verifiedBytes(modelFile(uri, true)) != null
          if (closed) promise.reject("E_NEEDLE_STALE", "The native module was closed.", null) else promise.resolve(verified)
        } catch (error: Exception) { promise.reject("E_NEEDLE_VERIFY", "Could not verify the private Needle model.", error) }
        finally { busy.set(false) }
      }
    }
    AsyncFunction("embedAsync") { uri: String, texts: List<String>, promise: Promise ->
      if (closed || !busy.compareAndSet(false, true)) promise.reject("E_NEEDLE_BUSY", "Needle is busy or unavailable.", null)
      else worker.execute {
        try {
          requireSupported()
          require(texts.size in 1..16) { "Needle accepts 1 to 16 texts." }
          val encoded = texts.map { text ->
            require(text.length <= 4096 && text.isNotBlank() && !text.contains('\u0000')) { "Needle needs nonempty text without NUL." }
            val buffer = Charsets.UTF_8.newEncoder().onMalformedInput(CodingErrorAction.REPORT).encode(CharBuffer.wrap(text))
            require(buffer.remaining() <= 4096) { "Needle text is too long." }
            ByteArray(buffer.remaining()).also { buffer.get(it) }
          }
          require(encoded.sumOf { it.size } <= 16384) { "Needle batch is too large." }
          val bytes = verifiedBytes(modelFile(uri, false)) ?: error("Download and verify the Needle model first.")
          val vectors = NeedleNative.embed(bytes, encoded.toTypedArray()).map { row -> row.map { it.toDouble() } }
          if (closed) promise.reject("E_NEEDLE_STALE", "The native module was closed.", null)
          else promise.resolve(mapOf("model" to NeedleNative.identity(), "dimensions" to vectors.first().size, "vectors" to vectors))
        } catch (error: Exception) { promise.reject("E_NEEDLE_EMBED", "Local retrieval could not finish: ${error.message}", error) }
        finally { busy.set(false) }
      }
    }
    OnDestroy { closed = true }
    // No native unload/cancel API exists. Pending callers discard stale results;
    // the process-global engine and its verified backing bytes live until exit.
  }
}

File: mobile/modules/keating-needle/ios/KeatingNeedleModule.swift
import CryptoKit
import ExpoModulesCore
import Foundation

public final class KeatingNeedleModule: Module {
  // Shared across module instances because the C engine owns process-global state.
  private static let worker = DispatchQueue(label: "keating.needle", qos: .utility)
  private static let admission = NSLock()
  private static var busy = false
  private let lifecycle = NSLock()
  private var closed = false
  private func current() -> Bool { lifecycle.lock(); defer { lifecycle.unlock() }; return !closed }
  private static func reserve() -> Bool { admission.lock(); defer { admission.unlock() }; if busy { return false }; busy = true; return true }
  private static func release() { admission.lock(); busy = false; admission.unlock() }
  private func directory() throws -> URL {
    guard NeedleBridge.supported() else { throw Exception(name: "E_NEEDLE_UNAVAILABLE", description: "Needle needs an ARM64 native build.") }
    var directory = try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      .appendingPathComponent("needle-retrieval", isDirectory: true)
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    var values = URLResourceValues(); values.isExcludedFromBackup = true; try directory.setResourceValues(values)
    return directory.resolvingSymlinksInPath()
  }
  private func modelFile(_ uri: String, partialAllowed: Bool) throws -> URL {
    guard let url = URL(string: uri), let components = URLComponents(url: url, resolvingAgainstBaseURL: false), url.isFileURL,
      (components.host ?? "").isEmpty, components.query == nil, components.fragment == nil else {
      throw Exception(name: "E_NEEDLE_PATH", description: "Needle needs its private local model file.")
    }
    let root = try directory().resolvingSymlinksInPath()
    let resolved = url.resolvingSymlinksInPath()
    guard resolved.deletingLastPathComponent().path == root.path,
      resolved.path == url.standardizedFileURL.path,
      resolved.lastPathComponent == NeedleBridge.filename() || (partialAllowed && resolved.lastPathComponent == NeedleBridge.filename() + ".part") else {
      throw Exception(name: "E_NEEDLE_PATH", description: "Needle model path is outside its private storage.")
    }
    return resolved
  }
  private func verifiedBytes(_ file: URL) throws -> Data? {
    let size = NeedleBridge.modelSize()
    guard let attributes = try? FileManager.default.attributesOfItem(atPath: file.path),
      attributes[.type] as? FileAttributeType == .typeRegular,
      (attributes[.size] as? NSNumber)?.intValue == size else { return nil }
    let handle = try FileHandle(forReadingFrom: file); defer { try? handle.close() }
    guard let data = try handle.read(upToCount: size + 1), data.count == size else { return nil }
    let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    return digest == NeedleBridge.modelSha() ? data : nil
  }
  public func definition() -> ModuleDefinition {
    Name("KeatingNeedle")
    Constants(["supported": NeedleBridge.supported(), "runtimeRevision": NeedleBridge.runtimeRevision(),
      "modelIdentity": NeedleBridge.modelIdentity().map { $0 as Any } ?? NSNull()])
    AsyncFunction("getModelDirectoryAsync") { () -> String in try self.directory().absoluteString }
    AsyncFunction("createModelFileAsync") { () -> String in
      let file = try self.directory().appendingPathComponent(NeedleBridge.filename() + ".part")
      guard file.resolvingSymlinksInPath().path == file.standardizedFileURL.path else { throw Exception(name: "E_NEEDLE_PATH", description: "Invalid Needle download path.") }
      if !FileManager.default.fileExists(atPath: file.path) {
        guard FileManager.default.createFile(atPath: file.path, contents: nil) else { throw Exception(name: "E_NEEDLE_FILE", description: "Could not create the model download.") }
      }
      return file.absoluteString
    }
    AsyncFunction("verifyModelFileAsync") { (uri: String, promise: Promise) in
      guard self.current(), Self.reserve() else { promise.reject("E_NEEDLE_BUSY", "Needle is busy or unavailable."); return }
      Self.worker.async {
        defer { Self.release() }
        do {
          let verified = try self.verifiedBytes(self.modelFile(uri, partialAllowed: true)) != nil
          if self.current() { promise.resolve(verified) } else { promise.reject("E_NEEDLE_STALE", "The native module was closed.") }
        } catch { promise.reject(error) }
      }
    }
    AsyncFunction("embedAsync") { (uri: String, texts: [String], promise: Promise) in
      guard self.current(), Self.reserve() else { promise.reject("E_NEEDLE_BUSY", "Needle is busy or unavailable."); return }
      Self.worker.async {
        defer { Self.release() }
        do {
          guard (1...16).contains(texts.count), texts.allSatisfy({ !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !$0.contains("\0") && $0.utf8.count <= 4096 }), texts.reduce(0, { $0 + $1.utf8.count }) <= 16384 else {
            throw Exception(name: "E_NEEDLE_INPUT", description: "Needle input exceeds the text batch limits.")
          }
          guard let bytes = try self.verifiedBytes(self.modelFile(uri, partialAllowed: false)) else { throw Exception(name: "E_NEEDLE_MODEL", description: "Download and verify the Needle model first.") }
          let result = try NeedleBridge.embed(model: bytes, texts: texts)
          if self.current() { promise.resolve(result) } else { promise.reject("E_NEEDLE_STALE", "The native module was closed.") }
        } catch { promise.reject(error) }
      }
    }
    OnDestroy { self.lifecycle.lock(); self.closed = true; self.lifecycle.unlock() }
    // No unload/cancellation API exists; callers discard stale results. The
    // engine and verified backing bytes remain alive until the process exits.
  }
}
